### PR TITLE
FW: Add CPU Frequency Manager

### PR DESCRIPTION
**Pull request by AthenasJimenez**
The end goal of this is adding an option that allows OpenDcdiag tool to change randomly CPU frequency during runtime.

NOTE: This is a WIP PR, so some things still don't work and some others are simply hard-coded.

Signed-off-by: Athenas Jimenez <athenas.jimenez.gonzalez@intel.com>

### CODE REVIEW

**Comment by AthenasJimenez** [comment]
@thiago and @lukasz-tuz any comments on this?

File: framework/sandstone.h [hunk 1]
```diff
@@ -387,6 +387,11 @@ struct test {
     /// free them in the test_cleanup function.
     void *data;
     struct test_data_per_thread *per_thread;
+
+    /// wheter frequency flags apply to this test or not.
```
Comment (lukasz-tuz) [inline framework/sandstone.h:391]: Perhaps we could make this an enum?

File: framework/sysdeps/linux/frequency_manager.hpp [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    FrequencyManager()
+    {
+        /* Discover supported and current frequencies */
+        std::filesystem::path cpuinfo_max_freq_p = "/sys/devices/system/cpu/cpu0/cpufreq/cpuinfo_max_freq";
+        max_freq_supported = get_frequency_from_file(cpuinfo_max_freq_p);
+
+        std::filesystem::path cpuinfo_min_freq_p = "/sys/devices/system/cpu/cpu0/cpufreq/cpuinfo_min_freq";
+        min_freq_supported = get_frequency_from_file(cpuinfo_min_freq_p);
+
+        std::filesystem::path scaling_max_freq_p = "/sys/devices/system/cpu/cpu0/cpufreq/scaling_max_freq";
```
Comment (lukasz-tuz) [inline framework/sysdeps/linux/frequency_manager.hpp:36]: For completeness of frequency control, FrequencyManager should also include support for uncore frequency scaling on supported processors (`/sys/devices/system/cpu/intel_uncore_frequency/`).

**Review by lukasz-tuz — APPROVED** [review]
Looks good, just a couple of comments which might as well be resolved in subsequent commits.

File: framework/sysdeps/linux/frequency_manager.hpp (hunk 2 above)
Comment (AthenasJimenez) [reply framework/sysdeps/linux/frequency_manager.hpp:36]: For uncore frequency, I'd like to leave it for a future commit. But sure, we definitely should do it.

File: framework/sysdeps/linux/frequency_manager.hpp [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+        for (int i=0; i< cpu_number; i++)
+        {
+            // Concatenate path
+            scaling_max_freq_p = DEFAULT_SYS_PATH;
+            scaling_max_freq_p += std::to_string(i);
+            scaling_max_freq_p += SCALING_MAX_FREQ_P;
+            
+            write_file(scaling_max_freq_p, max_freq_s);
+        }
```
Comment (lukasz-tuz) [inline framework/sysdeps/linux/frequency_manager.hpp:76]: I think it'd make more sense to make this method set frequency only for a single CPU and leave setting frequencies on other cores to the caller's discretion.

File: framework/sandstone.h (hunk 1 above)
Comment (lukasz-tuz) [reply framework/sandstone.h:391]: Ok, we obviously can't make this an enum :)

**Pull request closed without merging**